main/src/ca/uwaterloo/flix/language/phase/HtmlDocumentor.scala

```diff
@@ -0,0 +1,640 @@
+      }
+      sb.append("</span>")
+    }
+    sb.append("]</span>")
+  }
+
+  /**
+    * Document the given list of `FormalParam`s wrapped in `()`.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docFormalParams(fparams: List[TypedAst.FormalParam])(implicit flix: Flix, sb: StringBuilder): Unit = {
+    sb.append("<span class='fparams'>(")
+    docList(fparams.sortBy(_.loc)) { p =>
+      sb.append(s"<span><span>${esc(p.sym.text)}</span>: ")
+      docType(p.tpe)
+      sb.append("</span>")
+    }
+    sb.append(")</span>")
+  }
+
+  /**
+    * Document the given `Annotations`.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docAnnotations(anns: Ast.Annotations)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    for (a <- anns.annotations) {
+      sb.append(s"<code class='annotation'>${esc(a.toString)}</code>")
+    }
+  }
+
+  /**
+    * Document the given `SourceLocation`, `loc`, in the form of a link.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docSourceLocation(loc: SourceLocation)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    sb.append(s"<a class='source' target='_blank' href='${esc(createLink(loc))}'>Source</a>")
+  }
+
+  /**
+    * Document the the given `doc`, while parsing any markdown.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docDoc(doc: Ast.Doc)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    // DEFAULT_SAFE escapes HTML
+    val config = txtmark.Configuration.DEFAULT_SAFE
+    val parsed = txtmark.Processor.process(doc.text, config)
+
+    sb.append("<div class='doc'>")
+    sb.append(parsed)
+    sb.append("</div>")
+  }
+
+  /**
+    * Document the the given `Type`, `tpe`.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docType(tpe: Type)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    sb.append("<span class='type'>")
+    sb.append(esc(FormatType.formatType(tpe)))
+    sb.append("</span>")
+  }
+
+  /**
+    * Document the the given `Type`, `eff`, when it is known to be in effect position.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docEffectType(eff: Type)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    sb.append("<span class='effect'>")
+    sb.append(esc(FormatType.formatType(eff)))
+    sb.append("</span>")
+  }
+
+  /**
+    * Runs the given `docElt` on each element of `list`, separated by the string: ", " (comma + space)
+    */
+  private def docList[T](list: List[T])(docElt: T => Unit)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    for ((e, i) <- list.zipWithIndex) {
+      docElt(e)
+      if (i < list.length - 1) {
+        sb.append(", ")
+      }
+    }
+  }
+
+  private def clearOutputDirectory(): Unit = {
```

Comment:
Drop this-- in general Flix will never clear directories. (And if it will, it will be the job of the package manager and an eventual clean command.)

Reply:
Yes, I understand.